pandas/_libs/lib.pyx

```diff
@@ -1351,6 +1375,52 @@ cdef class Seen:
             self.datetime_ or self.datetimetz_ or self.nat_ or self.timedelta_
             or self.period_ or self.interval_ or self.numeric_ or self.object_
         )
+    cdef void get_curr_val(self, curr_seen_t *ret_val):
+        """
+        Gets the current seen val and casts it to ret_val's type.
+        Caller is responsible for making sure that the cast is safe
+        """
+        if self.curr_seen is CurrSeen.null_t:
+            if curr_seen_t is float64_t or curr_seen_t is complex128_t:
+                ret_val[0] = <curr_seen_t>NaN
```

Comment:
I really want to avoid casting like this - I think it will be really difficult to maintain this lack of type safety over time. It _might_ work better if the cast were contained within a class that also manages the type, but right now that is spread across a few objects

Reply:
Hm. I don't see a good way to do this without Tempita. Fused types don't work in cdef classes IIRC. 

Would raising an error if downcasting happens be OK?

(Not sure about allowing uint64 <-> int64 since that's slightly trickier, and is handled separately.)

Reply:
Really nice job getting rid of the casting. What happens if you remove it on this last one though? Not immediately obvious why this would still be necessary

Reply:
Should be able to get rid of that one too. thanks.